cfgov/v1/models/base.py

```diff
@@ -149,73 +149,74 @@ def generate_view_more_url(self, request):
-                        '-date_published'
-                    )
-                )
+        specific_categories = block.value['specific_categories']
+
+        for slug, should_get_related_posts in search_types:
+            if should_get_related_posts:
+                # Initialize Q query with a filter on this page's tags
+                query = models.Q(('tags__name__in', self.tags.names()))
+
+                # Get the children of the page
+                # e.g. all blog pages, newsroom pages, or events pages
+                parent = Page.objects.get(slug=slug)
+                query &= Page.objects.child_of_q(parent)
+
+                if slug == 'events':
+                    # We want to include archived events if dealing with events
+                    archive = Page.objects.get(slug='archive-past-events')
+                    query |= Page.objects.child_of_q(archive)
+                else:
+                    # Filter by provided categories for newsroom & blog
+                    if specific_categories:
+                        categories = ref.get_appropriate_categories(
+                            specific_categories=specific_categories,
+                            page_type=slug
+                        )
+                        if categories:
+                            query &= Q(('categories__name__in', categories))
+
+                final_query = AbstractFilterPage.objects.live().filter(
+                    query).distinct().exclude(
+                    id=self.id).order_by('-date_published')
                 # Apply similar logic as snippets.py's filter_by_tags method
                 # to enable AND filtering
                 if block.value['and_filtering']:
                     for tag in self.tags.names():
-                        type_query = type_query.filter(tags__name=tag)
-                related[search_type_name] = type_query[:block.value['limit']]
+                        final_query = final_query.filter(tags__name=tag)
+                related[slug.title()] = final_query[:block.value['limit']]
 
         # Return a dictionary of lists of each type when there's at least one
         # hit for that type.
         return {search_type: queryset for search_type, queryset in
                 related.items() if queryset}
 
+    def related_metadata_tags(self):
+        # Set the tags to correct data format
+        tags = {'links': []}
+        id, filter_page = self.get_filter_data()
+        relative_url = filter_page.relative_url(filter_page.get_site())
+        for tag in self.specific.tags.all():
+            tag_link = {'text': tag.name, 'url': ''}
+            if id is not None and filter_page is not None:
+                param = '?filter' + str(id) + '_topics=' + tag.slug
+                tag_link['url'] = relative_url + param
+            tags['links'].append(tag_link)
+        return tags
+
+    def get_filter_data(self):
+        for ancestor in self.get_ancestors().reverse().specific():
```

Comment:
Should `self.get_ancestors().reverse().specific()` go in its own variable before the loop? It looks like each method will get called each loop iteration.

Reply:
It's called in the outermost loop so it's only executed once